packages/syft/src/syft/custom_worker/config.py

```diff
@@ -0,0 +1,88 @@
+# stdlib
+from hashlib import sha256
+from pathlib import Path
+from typing import List
+
+# third party
+from packaging import version
+from pydantic import validator
+from typing_extensions import Self
+import yaml
+
+# relative
+from ..types.base import SyftBaseModel
+
+PYTHON_DEFAULT_VER = "3.11"
+PYTHON_MIN_VER = version.parse("3.10")
+PYTHON_MAX_VER = version.parse("3.12")
+
+
+class CustomBuildConfig(SyftBaseModel):
+    gpu: bool = False
+    # python_version: str = PYTHON_DEFAULT_VER
+    python_packages: List[str] = []
+    system_packages: List[str] = []
+    custom_cmds: List[str] = []
+
+    # @validator("python_version")
+    # def validate_python_version(cls, ver: str) -> str:
+    #     parsed_ver = version.parse(ver)
+
+    #     # TODO: Check if Wolfi OS/apk supports minor version of python
+    #     if parsed_ver.micro != 0:
+    #         raise ValueError("Provide only major.minor version of python")
+
+    #     if PYTHON_MIN_VER <= parsed_ver < PYTHON_MAX_VER:
+    #         return ver
+    #     else:
+    #         raise ValueError(
+    #             f"Python version must be between {PYTHON_MIN_VER} and {PYTHON_MAX_VER}"
+    #         )
+
+    @validator("python_packages")
+    def validate_python_packages(cls, pkgs: List[str]) -> List[str]:
+        for pkg in pkgs:
+            ver_parts = ()
+            name, ver = pkg.split("==")
+
+            if ver:
+                ver_parts = ver.split(".")
+
+            if not ver or len(ver_parts) <= 2:
+                raise ValueError(
+                    f'You must pin the package to an exact version. Got "{pkg}" expected "{name}==x.y.z"'
+                )
+
+        return pkgs
+
+    def merged_python_pkgs(self, sep=" ") -> str:
+        return sep.join(self.python_packages)
+
+    def merged_system_pkgs(self, sep=" ") -> str:
+        return sep.join(self.system_packages)
+
+    def merged_custom_cmds(self, sep=";") -> str:
+        return sep.join(self.custom_cmds)
+
+
+class CustomWorkerConfig(SyftBaseModel):
+    build: CustomBuildConfig
+    version: str = "1"
+
+    @classmethod
+    def from_dict(cls, config: dict) -> Self:
+        return cls(**config)
+
+    @classmethod
+    def from_str(cls, content: str) -> Self:
+        config = yaml.safe_load(content)
+        return cls.from_dict(config)
```

Comment:
I think we need a test for this class. I doubt it will work here since `CustomWorkerConfig` take a `CustomBuildConfig` argument, but `config = yaml.safe_load(content)` would be just a normal dictionary here so `cls(**config)` might not work. We probably need to use `cls.parse_obj(config)` instead.

Reply:
hmm `parse_obj` is not recursive either. This is something that is more convenient in Pydantic V2, we would have to do things a bit more manually here.

Reply:
ah my bad `parse_obj` actually works. there was a bug in my test.

Reply:
hmm from my test `cls(**config)` works too. that's nice though unexpected. will push the test.

Reply:
Yeah `ExampleObj(**some_dict)` is a [valid API shared in pydantic examples](https://docs.pydantic.dev/latest/#pydantic-examples)

Reply:
It's a bit different here. I'd expect `CustomWorkerConfig(**config)` to work if `config = {"version": version, "build": build_config}` if `build_config` is of type `CustomBuildConfig`. However here `config = yaml.safe_load(content)` so `build_config` would be a `dict`, not parsed into a `CustomBuildConfig` yet. It's a bit surprising to me that `CustomWorkerConfig(**config)` still works here.

Reply:
CustomWorkerConfig(**config) and CustomWorkerConfig.from_dict(config) are both valid syntax.

Reply:
I'm talking about the `from_str`

```py
def from_str(cls, content: str) -> Self:
    config = yaml.safe_load(content)
    return cls.from_dict(config)
```

which is the same as 


```py
def from_str(cls, content: str) -> Self:
    config = yaml.safe_load(content)
    return cls(**config)
```

If the yaml file looks like

```yaml
version: "1"
build:
  python_packages:
    - nltk==3.8.1
  ...
```

`config` would be {"version":  "1", "build": {"python_packages": ["nltk"==3.8.1], ...}} which surprises me that `cls(**config)` still works since I expected `cls(**config)` with `config={"version":  "1", "build": ConfigBuildConfig(python_packages=["python_packages==3.8.1"])}}` only.

Reply:
Anw I added more tests in #8288 for peace of mind.